uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-infinispan/src/main/java/org/uberfire/ext/metadata/backend/infinispan/provider/InfinispanContext.java

```diff
@@ -0,0 +1,218 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.uberfire.ext.metadata.backend.infinispan.provider;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.client.hotrod.marshall.ProtoStreamMarshaller;
+import org.infinispan.protostream.BaseMarshaller;
+import org.infinispan.protostream.SerializationContext;
+import org.infinispan.protostream.annotations.ProtoSchemaBuilder;
+import org.infinispan.query.remote.client.ProtobufMetadataManagerConstants;
+import org.uberfire.commons.lifecycle.Disposable;
+import org.uberfire.ext.metadata.backend.infinispan.exceptions.InfinispanException;
+import org.uberfire.ext.metadata.backend.infinispan.proto.KObjectMarshaller;
+import org.uberfire.ext.metadata.backend.infinispan.proto.schema.Schema;
+import org.uberfire.ext.metadata.backend.infinispan.proto.schema.SchemaGenerator;
+import org.uberfire.ext.metadata.backend.infinispan.utils.AttributesUtil;
+import org.uberfire.ext.metadata.model.KObject;
+
+import static java.util.stream.Collectors.toList;
+
+public class InfinispanContext implements Disposable {
+
+    private static final String PORT = "org.appformer.ext.metadata.infinispan.port";
+    private static final String HOST = "org.appformer.ext.metadata.infinispan.host";
+    private static final String USERNAME = "org.appformer.ext.metadata.infinispan.username";
+    private static final String PASSWORD = "org.appformer.ext.metadata.infinispan.password";
+
+    private static final String TYPES_CACHE = "types";
+    private static final String SCHEMAS_CACHE = "schemas";
+    private static final String PROTO_EXTENSION = ".proto";
+    private static final String SCHEMA_PROTO = "schema.proto";
+    private static final String ORG_KIE = "org.kie.";
+    private final RemoteCacheManager cacheManager;
+    private final KieProtostreamMarshaller marshaller = new KieProtostreamMarshaller();
+    private final SchemaGenerator schemaGenerator;
+
+    private final InfinispanConfiguration infinispanConfiguration;
+
+    private static InfinispanContext INSTANCE;
+
+    public static InfinispanContext getInstance() {
+        if (INSTANCE == null) {
```

Comment:
Probably should use [Lazy holder](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom) pattern to initialize the INSTANCE field.